### ISSUE TITLE
Support Platform Specific Extensions

### ISSUE THREAD

**Issue by windyGex** [microsoft/vscode#23251]
I can't find answer from stack overflow, so I open a issue.

**Comment by auchenberg** [comment]
In your extension activation callback, you are free to check and download extensions as soon as the extension get's activated. This is how Live Share is downloading their dependencies after installation
```
export async function activate(context: vscode.ExtensionContext) {
  /// 1 Check if deps are installed
 await downloadDependencies()
 // Update status bar or output channel to give an indicator of status
}
```

**Comment by windyGex** [comment]
If install dependencies when extension activate, I think it' too later and it maybe cost a long time.

**Comment by sandy081** [comment]
Idea is to provide a post install hook similar to [uninstall hook](https://github.com/Microsoft/vscode/issues/35006#issuecomment-367383767). This will be executed after an extension is installed or updated.

```json
"scripts": {
    "vscode:prepublish": "tsc -p ./",
	"vscode:install": "node ./out/src/extensionInstall"
    "vscode:uninstall": "node ./out/src/extensionUninstall"
},
```

An extension can install necessary libraries and stuff using this post install script. An unique location per extension is provided through an env variable for the extension to maintain any state. The same location will also be passed with post uninstall script.

**Comment by DonJayamanne** [comment]
> Idea is to provide a post install hook similar to uninstall hook. This will be executed after an extension is installed or updated.

I would prefer a better name. Today `vscode:prepublish` applies to developer of the extension, i.e. to extension authors. Having `vscode: install` would indicate it's related to extension authors again and not in end user machines.

**Comment by rchande** [comment]
@sandy081 Thanks for pointing this feature out to us on the C# extension. Our dependency downloading story is actually a little complicated. Each version of the extension ships with built in URLs to specific versions of binary dependencies, and, if the user has no options set, we'll download them on first run. However, we allow users to control what version of our dependencies they download. You can see this in action by playing with the option "omnisharp.path". If you:
* Set it to `latest`: When the extension starts, it will check online to determine the latest version of OmniSharp, download it if we haven't downloaded it before, and start that version. Each new version gets downloaded to its own folder.
* Set it to a specific version number: When starting OmniSharp, the extension will attempt to start it from a directory based on the version you specified. If it's not available, it will try to download it using a well known version -> url mapping. 

Basically, we need to be able to access our "unique location per extension" after install-time to read and write values. Would the new API support that?

**Comment by gregg-miskelly** [comment]
@sandy081 In addition to what Ravi said, I would first off like to say a thanks. I have been wanting this feature for so long. So I am very excited to see it happening!

One request: I would expect the most common action that an extension to want to do would be to install dependencies. However, with proxy servers, downloading files is not that easy of a task to get right. It would be great if there was some sort of 'DownloadFile' API 'vscode-DownloadFile' or npm package that was provided to extensions that would handle this task in the same way that VS Code does. In case it matters, as Ravi said, I think we would want some of our downloads to continue to go to the per-extension-version directory (because they change with essentially every version), while some of our downloads are stable enough that we might want to share them. So we would need to be able to specify the destination directory in such a way that either place could be targeted.

Here is the C# extension "download dependencies" code as a reference:
* Code to gather proxy settings: https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/NetworkSettings.ts
* Code to download files: https://github.com/OmniSharp/omnisharp-vscode/blob/master/src/packageManager/FileDownloader.ts#L44

**Comment by sandy081** [comment]
@DonJayamanne 

> I would prefer a better name. Today vscode:prepublish applies to developer of the extension, i.e. to extension authors. Having vscode: install would indicate it's related to extension authors again and not in end user machines.

`vscode:install` is related to extension author but it impacts end user machines at the end. For eg: downloading and installing libraries.

Going with `vscode:install` because we already have a post uninstall hook called `vscode:uninstall`.

**Comment by sandy081** [comment]
@rchande @gregg-miskelly 

Thanks for your feedback. Post install script will be given access to a unique location per extension (not per version). This location is provided to the script through process environment variable `VSCODE_EXTENSION_STORAGE_LOCATION`. You can update your post install script to download the dependencies matching the current version of extension being installed. You can use the above location to download and check the versions that are already downloaded. 

I do not think this script will be given access to configuration API. So if you have user configurable versions, then you can download the specific version during run time by checking in the above location.

> but we would need to somehow inform the user of progress since downloads can be somewhat large, at least for users with very slow internet connections.

Post install step runs as part of extension installation, so the extension installation progress in the extension view shows it. Logs show what step (eg: post install) is being running now.

> It would be great if there was some sort of 'DownloadFile' API or 'vscode-DownloadFile' npm package that was provided to extensions

I am completely aware of this and we are working to come up with a solution for this. Hence not including this here.


Let me know if this helps.

**Comment by gregg-miskelly** [comment]
> Post install script will be given access to a unique location per extension (not per version)

So you are saying that extensions will only have access to the per-extension location? If so, I would suggest changing the design to allow extensions to install to either place. Alternatively, if the 'DownloadFiles' API was really more a dependency management API where VS Code took care of removing outdated dependencies along with downloading the new ones, that could definitely work.

> Post install step runs as part of extension installation, so the extension installation progress in the extension view shows it. Logs show what step (eg: post install) is being running now.

Are you saying it will only show that it is on the 'post install' step and give no more granular progress than that? Or are you saying it will show that it is on the 'post install' step, and there will be some ability to see with more detail (ex: click on that to bring up an output window pane). If the former, I don't think that will be good enough unless the forthcoming 'DownloadFile' API solves this problem.

**Comment by gregg-miskelly** [comment]
BTW: I would like to throw out one other idea here. Instead of designing a bit of code that needs to run, really the missing feature here is that extensions can't tell VS Code about all of their platform-specific dependencies. So another way to solve this would be to make post-install hooks a feature that is normally implemented with a tiny bit of javascript that just returns (ex: by writing JSON to stdout) the set of packages that the extension needs. This way extensions don't need to all maintain a bunch of code to do this, VS Code can hopefully just reuse most of the existing extension download and extension file management code, and VS Code doesn't need to allow any of its normal APIs to work in these post-install hooks.

Example of how an extension might declare its dependencies:
```
{
  "dependencies": [
    {
      // The description would be used in the progress indicator
      "description": ".NET Core Debugger (macOS / x64)",
      // The id+version would be used to select the install directory and determine if the package was already downloaded
      "id": "vsdbg-ui",
      "version": "1.17.1",
      // URL which points to a .zip file (or other archieve formats if you want) to download from
      "url": "https://download.visualstudio.microsoft.com/download/pr/90e2038c-960d-4018-924e-30f520f887ab/2d53db027d1e67b899a7f083800b2bd4/coreclr-debug-osx-x64.zip",
      // The optional fallback URL is used if downloading from the primary URL fails
      "fallbackUrl": "https://vsdebugger.blob.core.windows.net/coreclr-debug-1-17-1/coreclr-debug-osx-x64.zip",
      // The optional hash is used to verify the file was correctly downloaded
      "sha512hash": "1234125215215..."
    },
    {
      "description": "OmniSharp for OSX",
      "id": "OmniSharp",
      "version": "1.32.6",
      "url": "https://download.visualstudio.microsoft.com/download/pr/a8501982-f057-4675-bf2f-57e4aa9f0990/f175b994e2c0a97ea7b874dafec7f303/omnisharp-osx-1.32.6.zip",
      "fallbackUrl": "https://omnisharpdownload.blob.core.windows.net/ext/omnisharp-osx-1.32.6.zip"
    }
  ]
}
```

**Comment by srivatsn** [comment]
+1 to @gregg-miskelly's idea above of making it declarative. An additional benefit of this could be that these dependencies could also be cached in the temp folder. That way, if the extension has a newer version but the dependencies don't change (like for LiveShare where we download .NET Core), then the update could be much faster as we'd just reuse the cached download on that machine.

**Comment by sandy081** [comment]
I do not agree letting the platform (VS Code) to do the dependency or resource management for the extensions is the right approach. It is hard for the platform to understand and provide an API that will fulfil all kinds of resource management requirements. Hence such an extension point will be incomplete always. So, I think extensions are the right ones to do their dependencies management and we provide a location where you can cache them and maintain the state.

I see that the dependencies approach is trying to solve following problems

1. **Showing Progress**

I agree as an an end user it is important to know 

- An extension is being installed
- A post installation process is going on
- Installation is finished or failed 

If user wants to know more about what is happening during post installation, user can see it in the logs. It's same with extension installation that the shared process log captures all details. 

One idea is that post install scripts logs their activity and VS Code show a status bar entry showing `Installing Extension C#...`. When user clicks on this entry, the installation log will be shown.

2. **Proxy issues**

I think this is a general issue and exposing the dependencies API will not help all cases. As said, we are working on it to provide a general solution that helps in all cases.

**Comment by avanderhoorn** [comment]
@sandy081 I think you statement above is true if you look at the 100% case, but there is a lot of value to be gained in solving the 80-95% case. I think you might find that the declarative API described by @gregg-miskelly might meet the majority of cases. Offline of this thread we have multiple discussions with @kieferrm and @egamma about this and I believe that the general consensus was that there would be value in providing something like this out of the box. I do agree that there are some underlying issues here (i.e. showing progress in a common way, proxy support, etc) but thats part of the reason that this issue exists. Lastly, to help get the declarative API to the point where it supports more cases, I don't see why there couldn't be strategic callbacks that people can tap to help direct things.

**Comment by sandy081** [comment]
@avanderhoorn If I am not wrong the reason for downloadable resources extension point proposal was to overcome the proxy issue otherwise the post install script extension point is the better way for the extensions to download / manage resources. Also, this resources extension point is not a complete and generic solution as this will not help extensions if they want to download during run time.

Again, we are doing some investigation and research on how to provide a generic solution for proxy issues. If that yields positive results the resources extension point might be redundant.

**Comment by sandy081** [comment]
Tagging for `api-proposal` because this needs to expose `globalStorage` api in `vscode.ExtensionContext`.

**Comment by rchande** [comment]
@sandy081 What's the rationale for providing a post-install script file as opposed to a javascript callback (as @gregg-miskelly suggested)? If extensions provide a post-install callback that calls a standard VS Code downloading API, you can make progress reporting very consistent.

**Comment by sandy081** [comment]
@rchande post-install script is going to be a node script that gets executed after installing the extension. VS Code API is available only for running extensions and hence the newly installed extension does not has access to it as it is not running. A new API has to be defined if post-install script wants to show some UI elements which we want to avoid.

I am consolidating the proposal here

- An extension can participate in its post installing step by providing the node script entry point using `vscode:install` key in `scripts` section of `package.json`
- VS Code shows a status entry for post install step and when user

**Comment by avanderhoorn** [comment]
@sandy081 I think the above is a step forward and an improvement but I still think there is a wider problem here and that is the general requirement that more and more extensions are having of downloading dependencies post installation. 

Best practice says that extensions should be kept small and then any external dependencies downloaded post installation. This is fine, but it puts a lot on extension authors to have to reinvent the wheel every time (i.e. each extension needs to reauthor this code). The worst part is that its surprisingly hard to get this code right in a robust way and it turns out that it fails a lot more often that most extension authors would expect and just don't know since they don't that level of insights setup.

Hence I really thing its worth exploring if there is a more general solution here. As mentioned previously I think there is some basic thing that VS Code could do here which would improve the experience and reliability for all involved. 

As you can see this isn't just about proxies, its about the winder user and extension author experience.

**Comment by sandy081** [comment]
@avanderhoorn I got your argument but here are the points why I am not in favour of having resource management 

- We would like to avoid VS Code having non-editor specific API or functionality.

- As I said before there is no generic way that a platform can support for downloading and managing resources that will suffice all extensions or use cases that are there outside. 

- I see this (resource management) as an util functionality that extensions could  reuse rather than a VS Code editor feature. A node library can suffice the reuse purpose. Even though not completely related, we followed the same for language server libraries.

From a platform perspective, it is important to provide proper entry points to contribute their functionality.

**Comment by sandy081** [comment]
@alexandrudima FYI.

**Comment by jrieken** [comment]
> There is already storagePath provided to extensions to store workspace specific state. This will allow extensions to store global state for eg., downloading resources

We will clean up how?

**Comment by gregg-miskelly** [comment]
> there is no generic way that a platform can support for downloading and managing resources that will suffice all extensions or use cases that are there outside.

I personally think that a basic 'dependency manager' could be designed that would meet the primary needs of something very close to 100% of extensions. It certainly couldn't fill every scenario, for example the C# scenario of downloading a different OmniSharp than that which was shipped. But there are still activation hooks as a fall back, and if there was a golden path I bet you would find that all the extension authors would happily move to it instead of needing to maintain some very tricky code.

As you suggested, this certainly could be done as a node package rather than something built into VS Code itself. But I would worry that doing that will make this problem harder to solve since you can no longer rely on VS Code's infrastructure (existing extension management and download support) and it can longer easily integrate with the VS Code UI (ex: showing progress for downloads, telemetry for downloads).

**Comment by alexandrudima** [comment]
@sandy081 An alternative would be to add support for dynamic dependencies (at least based on platform).

For example, there could be three extensions: 
* `ms-vscode.csharp-deps-win32`, 
* `ms-vscode.csharp-deps-darwin` and 
* `ms-vscode.csharp-deps-linux`

Each one of these can be published independently (you could think of them like library extensions). They would only contain what is today downloaded independently.

Then, the main extension `ms-vscode.csharp` defines in `extensionDependencies` that it wants `ms-vscode.csharp-deps-${platform}`. When installing the main extension, we can resolve `${platform}` and download/install the correct one. The main extension will only get activated if the dependency is on disk, enabled and activated.

IMHO this has the advantage of reusing existing concepts and makes installing an extension talk only to our marketplace. For example, I know we use a special CDN for faster download speeds in China, and perhaps not all extension authors will or could do that sort of thing for platform specific binaries.

**Comment by jrieken** [comment]
Ideas @sandy081 and I have discussed is to keep the separate entry point but to execute that in the extension host with access to the `vscode`-module. That means can have a script which is different from the extension entry but which runs in the same context so code sharing works and API is available. Obviously not all API makes sense to be used in there (e.g. a post install hook shouldn't register language providers etc) but no harm can be done (which cannot be done else how)

**Comment by alexandrudima** [comment]
> [...] to keep the separate entry point but to execute that in the extension host 
 with access to the `vscode`-module. That means can have a script which is different from the extension entry but which runs in the same context so code sharing works and API is available.

But then what is the purpose of such a script, since this script can simply be inside the `activate` method (which is what extensions do today). i.e. That is why the `activate` method can return a `Promise`:

```
export async function activate() {
   await downloadMyStuff();
   // ...
}
```

**Comment by jrieken** [comment]
> But then what is the purpose of such a script,

Yeah, if we would call activate right after install this wouldn't be an issue IMO.

**Comment by sandy081** [comment]
But activating the extension immediately after installing is like running the extension when extension host is started. It means we have to support hot loading of an extension which I am always in favour of.

**Comment by sandy081** [comment]
Deferring this as the solution that is agreed upon is too big to implement now.

**Comment by tboby** [comment]
For context, I have a single extension which bundles a compiled .NET core console app targeting all three platforms. This is the language server, which isn't published anywhere else. Perhaps it should be an external dependency already, but that adds the complication of file hosting.

@alexandrudima 
> Each one of these can be published independently (you could think of them like library extensions). They would only contain what is today downloaded independently.

I see two cases where you need this:
1. When you have platform dependent binaries that are tightly coupled to your extension. (My case)
2. When you have platform dependent binaries that are published independently of your extension (E.g. omnisharp?)

Re 1:
**Staying in sync**: 
One downside of this is that it adds the opportunity for weird situations if publishing fails for one.

My process would become:
1. Commit changes to language server (.NET core console app)
2. Build and publish `tboby.cwtools-vscode-win32-0.5`
3. Build and publish `tboby.cwtools-vscode-darwin-0.5`
4. Build and publish `tboby.cwtools-vscode-linux-0.5`
5. Build and publish `tboby.cwtools-vscode-0.5`

What happens if I fail at step 3? How do I recover?
- I can't republish `win32-0.5`, so I can't rerun my whole process. Do I have to write a build script to handle that and skip?
- What version will users of `tboby.cwtools-vscode-0.4` now receive? Do they get `win32-0.5` on windows and `linux-0.4` on linux?

**Dev workflow:**
Splitting one extension into several tightly coupled ones has workflow problems. Currently I have a single git submodule dependency on the language server code.
- Is it possible to debug a group of extensions together?
- How would I test a new version of `tboby.cwtools-vscode-win32` in `tboby.cwtools-vscode` without publishing it?

Re 2:
**Binaries:**
If the binaries are published independently of your extension (e.g. a language server), bundling them into an extension doesn't add much benefit for a lot of cost. If I don't control the language server publishing, I then need to remember to republish that extension every time a new version releases.

Sorry if this is too much feedback too early, but I took the title change and addition to roadmap as a cue :)

**Comment by sandy081** [comment]
@tboby Thanks for your input. There is a new proposal (in the line of platform specific)  which we thought about and also discussing with our Marketplace team that hosts extensions.

We would like Marketplace to support Platform specific extensions. What it means is that if an extension has platform dependencies then extension author can bundle the same extension version for all platforms and publish it to the Marketplace. We do not have concrete approach for such a publishing step but you might have a single command to publish a version of extension with different platform flavours. VS Code then installs extension based on the user platform. 

It is up to extension author to extract the library as a separate extension so that other extension can depend on or bundle everything into a single extension. 

So in your case, you can just say build and publish `tbody.cwtools-vscode` with some platform specific arguments (tbd) and vsce can publish the same version of your extension in different platforms. 

This is a proposal which we are discussing with Marketplace team to support.

**Comment by DonJayamanne** [comment]
@sandy081

Doesn't this complicate downloading VSIX for offline installs.
I.e. when downloading a VSIX, we now have to download platform specific VSIX, i.e. multiple VSIX if we plan on using it for offline installs

**Comment by sandy081** [comment]
Good point. A VSIX can be self contained with all platform flavours. VS Code can pick the right flavour from VSIX while installing.

**Comment by tboby** [comment]
Adding a single action to `vsce` that does the bundling and publishing together then having VS Code install the appropriate version is a good idea, solves all my concerns.

Small drawback to having VS Code download the bundle (not just the platform specific part) is that at least 50% of the download is wasted.

**Comment by sandy081** [comment]
> Small drawback to having VS Code download the bundle (not just the platform specific part) is that at least 50% of the download is wasted.

VS Code can directly ask Marketplace for the version of the extension for the given platform while installing extension from VS Code UI. But when users wants to download a VSIX directly from Marketplace or package the VSIX using VSCE then they can have self contained VSIX which can also be handled by VS Code.

**Comment by Vyshakh18** [comment]
Hi,
I was developing an extension which uses native module, which requires platform specific treatment - and it's after publishing the extension, i hit the hard reality of this not being supported as of now.
So just to know - do you have any rough estimate of when this feature is going to be supported, atleast via the Marketplace route being discussed here?

**Comment by sandy081** [comment]
@Vyshakh18 Planning to bring this in our June milestone plan.

**Comment by sean-mcmanus** [comment]
@sandy081 Is this still planned for the July milestone or did it get delayed?

**Comment by sandy081** [comment]
@sean-mcmanus Not for july. But the proposal was discussed with Marketplace team and there is a movement. I will bring this up in our planning in next month.

**Comment by FerhatAkgun** [comment]
Are there any updates or concrete milestone plans?

**Comment by NLueg** [comment]
@sandy081 Hi, is there any update on this? I would appreciate this feature very much. Is there anything the community can do to bring this forward?

**Comment by sandy081** [comment]
We are working with Marketplace to get a proper solution.

**Comment by prashantvc** [comment]
Hey Everyone!

I am Prashant, PM on the VS Marketplace team. The team serving extensions to VS Code, we are in the process of prioritising the VS Marketplace backlog, I would love to get fresh perspective on this issue

 We can set up a quick call, you can pick the meeting time that works for you here: https://aka.ms/vsm-call

You can contact me at prchol@microsoft.com if you prefer emails.

**Comment by felipecassiors** [comment]
To have platform-specific extensions certainly solves most of the problem, since we could simply package all the needed binaries into the `.vsix`.

One thing to note is: some licenses don't allow binary redistribution, which in this case we would still need a mechanism for downloading the dependencies.

**Comment by gjsjohnmurray** [comment]
@felipecrs made this good point:
> One thing to note is: some licenses don't allow binary redistribution, which in this case we would still need a mechanism for downloading the dependencies.

@joaomoreno @fiveisprime it seems to me that the solution to the whole issue cannot just lie with Marketplace, so the `vscode:install` script proposal from @sandy081 [here](https://github.com/microsoft/vscode/issues/23251#issuecomment-441962837) is going to be necessary anyway. And it'll help us all while we wait for Marketplace changes to come through someday.

**Comment by davidanthoff** [comment]
Just checking in again what the status of this is :) We (the Julia language extension) would love to see the platform specific extension story implemented, that would be the most convenient solution for us.

**Comment by aallrd** [comment]
Hello,
Do you have an update on this feature?
We have to manually download the full extension and pull it from an internal server to install it offline.

**Comment by prashantvc** [comment]
Unfortunately, we do not have any update to share now.
But we have something in works, appreciate your patience

**Comment by felipecrs** [comment]
I think something is about to emerge... 

https://github.com/microsoft/vscode-vsce/commit/5b4795340602a97b16397328f71bc77e54f2bd6a

**Comment by prashantvc** [comment]
> I think something is about to emerge...

it is too early to use this feature for extensions, I will keep you all informed once we have something concreate to share.
I would love to get early feedback, please contact me; details are on my profile :)

**Comment by isidorn** [comment]
@gregg-miskelly just a heads up that we will soon have something for try out. Will you have time to give this a try on the C# extension side? We are very much looking for early feedback. I will ping here again once we push the initial support for this with some docs and samples.

@wardengnaw @sean-mcmanus who would be the best on the C++ extension side to ping for platform specific extension structure?

**Comment by felipecrs** [comment]
If it helps, I can also help to test at the [ShellCheck extension](https://github.com/vscode-shellcheck/vscode-shellcheck) side.

**Comment by isidorn** [comment]
All feedback will be welcome of course. I'll post here once we have updates. Thanks!

**Comment by sean-mcmanus** [comment]
@isidorn Maybe @Colengms

**Comment by prashantvc** [comment]
I am looking for extensions authors to test the platform specific extensions publishing workflow.
If you want to participate, please fill up this form https://forms.office.com/r/M4hgYqR6fx

**Comment by felipecrs** [comment]
@prashantvc I don't have permission to fill the form, it says it needs a business account instead of my Microsoft account.

**Comment by prashantvc** [comment]
> @prashantvc I don't have permission to fill the form, it says it needs a business account instead of my Microsoft account.

Sorry about it, let me fix it

**Comment by isidorn** [comment]
We now support the full flow for platform specific extensions 🚀  
This docs have all the details:
https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions

@testforstephen @jdneo @nickzhums @aeschli @colengms @gregg-miskelly @brettcannon @davster @gjsjohnmurray  and interested members of our extension community: please try transforming your extensions into platform specific extensions and let us know how it goes. We are looking forward to you feedback, so for now let's keep this issue open for discussions and that we can react with improvements. Thanks 👏 

Thanks a lot to our friends in the Marketplace team for enabling this @Saikanth007 @pratikgawali @prashantvc and the rest of the team 👏

**Comment by iliazeus** [comment]
@isidorn the wording isn't quite clear in the docs: what happens with previous extension versions when I first publish a platform-specific version? Are they still available for all platforms?

**Comment by isidorn** [comment]
@iliazeus good question, I can clarify that in the next version of the docs.
Once you release a platform specific `vsix` for any platform your extension becomes a `platform specific extension`, thus it is only supported for the platforms for which you have uploaded packages for. So your older versions will not be available for install for all platforms.

**Comment by Eskibear** [comment]
For Java extension, I want to pack platform-specifc JDKs. I want to pack JDKs for some platforms (e.g win32-x64/darwin-x64/linux-x64), and publish platform-specific extensions. For uncoverred platforms (like win32-ia32), we publish non-platform-specific extension without packing any JDK, as a fallback.

To verify if above proposal works for the moment, I tested with a [sample extension](https://github.com/Eskibear/vscode-extension-sample). 
I use vsce@1.99.0 to package vsix files, and marketplace portal page to publish/update. 
.vsix files I published: [yanzh.sample.zip](https://github.com/microsoft/vscode/files/7168627/yanzh.sample.zip)

I published them in below order, testing which distribution my PC(win32-x64) and Macbook(darwin-x64) can get.
(everytime after uninstalling the extension, I manually removed corresponding folder under `~/.vscode-insiders/extensions` to avoid being affected by cache)

<body link="#0563C1" vlink="#954F72">

After publishing | PC (win32-x64) gets | Macbook Pro(darwin-x64) gets
-- | -- | --
0.0.1 | 0.0.1 | 0.0.1
0.0.2-win32-x64 | 0.0.2-win32-x64 | 0.0.1
0.0.2 | **0.0.2 <--Oops** | 0.0.2
0.0.2-darwin-x64 | 0.0.2 | 0.0.2-darwin-x64

</body>

Overall it's working well. But after I published 0.0.2, my PC fetched 0.0.2 from marketplace although there was a "more matching" one (0.0.2-win32-x64). I'm expecting it to fetch 0.0.2-win32-x64, otherwise I have to always publish the "generic" one before platform-specific ones.

**Comment by tboby** [comment]
@isidorn Congrats on getting this done, being looking forward to this feature!
A few questions that aren't clear to me from the documentation:

1. Which release is the "main" one:
    i. Which vsix shows as the version on the marketplace?
ii. Which vsix determines the readme page on the marketplace and in the extension tab of VS Code?
2. From `1.61.0` will VS Code *never* use a non-targeted VSIX if I have published *any* targeted ones?
3. What happens if your versions get out of sync? 
 i. Does VS `1.61.0+` treat each platform as a separate extension and simple aggregate stats on the marketplace side of things?
ii. If I release 2.0 for one platform, does this in any way impact the other platforms?

**Comment by testforstephen** [comment]
I'm trying to figure out how the universal and platform versions can coexist, so i used `vsce@1.99` tool to publish multiple versions for a test extension. 

<body link="#0563C1" vlink="#954F72">

vsce publishing | Status | Details
-- | -- | --
~~v1.15.0: `vsce publish --target win32-x64`~~ | ~~FAIL~~ | ERROR  Platform specific extension is supported by VS Code >=1.61. Current 'engines.vscode' is '^1.53.2'. And updating engine version to 1.61 fixes the error.
v1.15.0: `vsce publish --target win32-x64` | PASS |
v1.15.0: `vsce publish --target darwin-x64` | PASS |
_v1.15.1_: `vsce publish`                               |     PASS |
_v1.15.1_: `vsce publish --target darwin-x64`  | PASS |
**v1.15.2**: `vsce publish --target darwin-x64`  | PASS |
**v1.15.2**: `vsce publish`                                     | FAIL |   ERROR  publisher_id.vscode-platform-specific-sample v1.15.2 already exists. Version number cannot be the same.
</body>

**Issue 1:**
First use "vsce publish", then "vsce publish --target platform", and it works. But change the order of these two commands and it doesn't work. This should be a bug.

**Issue 2:**
After publishing, I used different OS to verify how the client displays these versions.
In windows 1.61-insider, it shows the highest compatible version as 1.51.1. And I can also install another version like 1.51.0. this is expected.
In macOS 1.61-insider, the highest version it shows is 1.52.0. This is also expected. However, the "Install another version..." in the Extension view " will show 4 versions, as shown below. It shows two versions of 1.51.1. In this case, it is better to attach the platform tag or directly hide the universal one if the same platform one exists.

![image](https://user-images.githubusercontent.com/14052197/133422922-17a9660a-cb0b-4d79-9ada-d4140182c061.png)

**Comment by prashantvc** [comment]
I will take a stab at it. We are working on documenting these FAQs :)

> 1. Which release is the "main" one:
>    i. Which vsix shows as the version on the marketplace?

Latest version of the latest upload is the "main" release. 

>    ii. Which vsix determines the readme page on the marketplace and in the extension tab of VS Code?

Again, latest version of the latest upload. This is the platform limitation of the Marketplace; we will fix this in the future to show appropriate details in the marketplace based on the platform

> 2. From `1.61.0` will VS Code _never_ use a non-targeted VSIX if I have published _any_ targeted ones?

If VS Code finds a compatible targeted extension it will use it, otherwise it will fallback to non-targeted or last compatible version of an extension

> 3. What happens if your versions get out of sync?

You will always see details from the latest version of latest upload. But VS Code will download the compatible VSIX even it's the older version of an extensions

>    i. Does VS `1.61.0+` treat each platform as a separate extension and simple aggregate stats on the marketplace side of things?

Yes, right now we will show aggregate stats in the marketplace, but it will improve in the future.

>    ii. If I release 2.0 for one platform, does this in any way impact the other platforms?

No, You will always see details from the latest version of latest upload. But VS Code will download the compatible VSIX even it's the older version of an extensions

**Comment by prashantvc** [comment]
@testforstephen 

 > I'm trying to figure out how the universal and platform versions can coexist, so i used vsce@1.99 tool to publish multiple versions for a test extension

No, the universal and platform versions _cannot_ coexist for the same version of an extension

> Issue 1

Is this for same version of an extension? Can you give an example, please?

> Issue 2

Yeah this looks like a bug. Could you file it on `vsce` repo?

**Comment by isidorn** [comment]
@tboby @testforstephen @eskibear thanks for giving this a try and for providing feedback 👏 
I believe Prashant answered the majority of the questions.

We do not recommend to have platform specific vsix and universal vsix for the same version. Instead we recommend that for a version an extension would publish a platform specific package for each platform (even if vsix is shared across different platforms). If this is too much manual work let us know (we are already looking into improve the flow in `vsce` for this). 

Only @Eskibear issue was not covered by Prashat: @sandy081 is working on a fix and will probably push a fix for this soon.

**Comment by gregg-miskelly** [comment]
Thanks so much for implementing this. I am very excited!

Two questions:
1. Is there any limit on the size of a .vsix that we need to be aware of?
2. If we use platform-specific .vsix's for our official releases, but still use platform-agnostic .vsix's for beta releases (which would have a higher version number) is that going to be a problem?

**Comment by felipecrs** [comment]
> If this is too much manual work let us know (we are already looking into improve the flow in `vsce` for this).

I think it's the case. For example, I use [Semantic Release](https://github.com/semantic-release/semantic-release) to manage the versioning of the [ShellCheck extension](https://github.com/vscode-shellcheck/vscode-shellcheck) through [semantic-release-vsce](https://github.com/felipecrs/semantic-release-vsce). I cannot even think of a way of handling two versions at once for the platform specific and not platform specific extension with Semantic Release.

**Comment by sandy081** [comment]
Thanks for feedback. Fixed following:
- VS Code will now install the most suitable vsix 
- Fixed the version duplication

**Comment by isidorn** [comment]
@gregg-miskelly 
1. Current limit per platform package is `200mb`. This is not set in stone, we can increase if we hear from you it is needed.
2. This is not going to be a problem. We always take the highest version which is compatible. However please be aware that we are looking into improving the extension insider flow and we have a [proposal here](https://github.com/microsoft/vscode/issues/15756#issuecomment-916728348). This would work very well with Platform Specific Extension structure

@felipecrs this is why we recommend automating this flow. Though we will also try to improve the manual flow as well. I will keep you updated. Thanks for feedback.

**Comment by jakebailey** [comment]
I would also think it would be a good idea to offer up a "generic" platform that's a fallback if you haven't published the right one. I don't think it's good for extensions to have to keep track of every platform VS Code could support (present and future) and ensure they publish versions for each of them. There's always the case that a new platform (say, darwin arm64) appears, then some older extension can no longer be used because it won't be installed unless you have the platform available.

But, I know this (and the insiders versioning) is sort of limited by what the marketplace can do.

**Comment by isidorn** [comment]
@jakebailey this is a good suggestion, however we are a bit limited due to marketplace support.
We should keep this in mind and potentially improve later, but this is not in the cards for the near future.

**Comment by gregg-miskelly** [comment]
> 1. Current limit per platform package is 200mb. This is not set in stone, we can increase if we hear from you it is needed.

Thanks. For the C# extension, have one platform (darwin-arm64) where we are fairly close to the limit -- 198425276 bytes (189.2 MB) in my quick test. I can do some experiments to see if we can reduce this a bit. But we have have an ask to raise this slightly.

**Comment by isidorn** [comment]
@gregg-miskelly ok, so it seems like we are good for now but I have notified the Marketplace that we would like to increase the size limit. Once they do that I will let you know.

**Comment by isidorn** [comment]
As I mentioned above
We now support the full flow for platform specific extensions 🚀
This document has all the details:
https://code.visualstudio.com/api/working-with-extensions/publishing-extension#platformspecific-extensions

Thus I am closing this issue. If there are questions feel free to ask, we will continue monitoring this issue even though it is closed. Thank you very much 👏

**Comment by isidorn** [comment]
@felipecrs yes you can, but you need to publish your `amd64` vsix for `amd64`, and your general one for each other platform. This is an issue to help make this flow smoother https://github.com/microsoft/vscode-vsce/issues/625

**Comment by valentjn** [comment]
@isidorn When Marketplace increases the size limit, is this announced in this issue or in another? Or is this documented somewhere? I can't find the info about the 200MB anywhere else than here (the previous limit of 100MB is also hard to find, it's buried in some old release notes of VS Code).

Whether it's 200MB or say 250MB or 300MB makes a big difference for my extension with respect to migrating to platform-specific extensions, but I don't know a way of finding out the current limit other than trying to publish and see what happens. Obviously, at that point, a lot of effort has been invested, in vain if the limit is lower than expected.
